lib/compiler/aro_translate_c.zig

```diff
@@ -215,80 +221,251 @@ fn prepopulateGlobalNameTable(c: *Context) !void {
+    var cur_name = want_name;
+
+    if (!c.weak_global_names.contains(want_name)) {
+        // This type wasn't noticed by the name detection pass, so nothing has been treating this as
+        // a weak global name. We must mangle it to avoid conflicts with locals.
+        cur_name = try std.fmt.allocPrint(c.arena, "{s}_{d}", .{ want_name, c.getMangle() });
+    }
+
+    while (c.global_names.contains(cur_name)) {
+        cur_name = try std.fmt.allocPrint(c.arena, "{s}_{d}", .{ want_name, c.getMangle() });
+    }
+    return cur_name;
+}
+
+fn transRecordDecl(c: *Context, scope: *Scope, record_node: NodeIndex, field_nodes: []const NodeIndex) Error!void {
+    const node_types = c.tree.nodes.items(.ty);
+    const raw_record_ty = node_types[@intFromEnum(record_node)];
+    const ty = raw_record_ty.canonicalize(.preserve_quals);
+    const record_decl = ty.data.record;
+    if (c.decl_table.get(@intFromPtr(record_decl))) |_|
+        return; // Avoid processing this decl twice
+    const toplevel = scope.id == .root;
+    const bs: *Scope.Block = if (!toplevel) try scope.findBlockScope(c) else undefined;
+
+    var is_union = false;
+    var container_kind_name: []const u8 = undefined;
+    var bare_name: []const u8 = c.mapper.lookup(record_decl.name);
+
+    if (ty.is(.@"union")) {
+        container_kind_name = "union";
+        is_union = true;
+    } else if (ty.is(.@"struct")) {
+        container_kind_name = "struct";
+    } else {
+        try c.decl_table.putNoClobber(c.gpa, @intFromPtr(record_decl), bare_name);
+        return failDecl(c, 0, bare_name, "record {s} is not a struct or union", .{bare_name});
+    }
+
+    var is_unnamed = false;
+    var name = bare_name;
+    if (c.unnamed_typedefs.get(@intFromPtr(record_decl))) |typedef_name| {
+        bare_name = typedef_name;
+        name = typedef_name;
+    } else {
+        // Record declarations such as `struct {...} x` have no name but they're not
+        // anonymous hence here isAnonymousStructOrUnion is not needed
+        if (ty.isAnonymousRecord(c.comp)) {
+            bare_name = try std.fmt.allocPrint(c.arena, "unnamed_{d}", .{c.getMangle()});
+            is_unnamed = true;
+        }
+        name = try std.fmt.allocPrint(c.arena, "{s}_{s}", .{ container_kind_name, bare_name });
+        if (toplevel and !is_unnamed) {
+            name = try mangleWeakGlobalName(c, name);
+        }
+    }
+    if (!toplevel) name = try bs.makeMangledName(c, name);
+    try c.decl_table.putNoClobber(c.gpa, @intFromPtr(record_decl), name);
+
+    const is_pub = toplevel and !is_unnamed;
+    const init_node = blk: {
+        var fields = try std.ArrayList(ast.Payload.Record.Field).initCapacity(c.gpa, record_decl.fields.len);
+        defer fields.deinit();
+
+        var functions = std.ArrayList(ZigNode).init(c.gpa);
+        defer functions.deinit();
+
+        var unnamed_field_count: u32 = 0;
+
+        for (record_decl.fields, field_nodes, 0..) |field, field_node, i| {
```

Comment:
`record_decl.fields` and `field_nodes` won't necessarily have the same length; the former is the fields; the latter in this case is all the nodes inside that struct, which might include e.g. anonymous record decls. I think you'll need 2 loops, one over the nodes and one over the fields. Any nodes that are decls should be translated. For the fields loop, you can use `field.ty` for the type instead of looking it up in the node types slice. I also don't think you have to canonicalize the field type.

Reply:
Ah, that makes sense. Thanks for explaining that, I'll update that.